### ISSUE TITLE
Apply middlewares once.

### ISSUE THREAD

**Issue by SebastianStehle** [graphql-dotnet/graphql-dotnet#3337]
## Summary

Change the logic to apply the middleware only once.

### Basic example

I have the special requirements to share graphql types between schemas. I have described in detail here: https://github.com/graphql-dotnet/graphql-dotnet/issues/3150

Thererfore I cannot use middlewares, because they would be applied several times.

I would like to change this behavior. I have two ideas:

## Change the logic a little bit

1. Use a special Middleware resolver and check if the resolver is of this type, e.g. we could check for the type on the following line:

https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Collections/SchemaTypes.cs#L382

Something like

```
foreach (var field in complex.Fields.List)
{
	// Middleware has already been applied.
	if (field.Resolver is MiddlewareResolver)
	{
		continue;
	}

	var inner = field.Resolver ?? NameFieldResolver.Instance;

	var fieldMiddlewareDelegate = transform(context => inner.ResolveAsync(context));

	field.Resolver = new MiddlewareResolver(fieldMiddlewareDelegate.Invoke);
}
```

## Change the signature of the `IFieldMiddlewareBuilder`

If the method would be something like the following:

```
void Apply(Field field);
```

I could write my custom middleware builder.

### Motivation

See above. But the code for middlewares is super easy, so that I could also write my own. Not a showstopper.

**Comment by Shane32** [comment]
As of now, we don't have any plans to support sharing non-scalar graph types between schemas.  We do already have code that protects graph types from being initialized multiple times, which indirectly prevents registering middleware twice for the same fields (see #3248 ).  You are likely aware of this since it would need to be disabled for your use case.

I suggest registering a factory method within your DI engine for your schema class/interface, calling `schema.Initialize()` and then iterating over the graph types and applying the middleware manually.

For example:

```cs
services.AddSingleton<ISchema>(provider => {
    var schema = new MySchema(provider);
    schema.Initialize();
    // pull registered middleware from DI here ???
    foreach (var type in schema.AllTypes) {
        if (type is IObjectGraphType objGraphType) {
            // loop through fields and check if middleware is applied, and if not, apply it
        }
    }
}
```

There might be a better way, potentially re-using more of the middleware code that exists within GraphQL.NET, but this is just what comes to mind.

**Comment by SebastianStehle** [comment]
Yes, I was "afraid" you would say that ;) ... But it is not that much code, so it is fine for me.

**Comment by Shane32** [comment]
At one time I had worked on code to make the interfaces all read-only, so the classes were mutable, and the interfaces were read-only.  Conceptually, upon initializing the schema you would have a immutable schema instance.  See some PRs linked below.  It ended up that there was too drastic of changes required with little perceived benefit in practical usefulness.  I still think it's generally the correct concept, but it would require further thought and redesign to implement.

- #1578
- #1572
- #1590
- #1593
- #1594

**Comment by Shane32** [comment]
Eventually perhaps we have a `SchemaBuilder` class with a `Build` method that builds the immutable `ISchema` instance.  But at present, I am not working on those types of changes.

**Comment by sungam3r** [comment]
@Shane32 I suggest to close as wontfix.

**Comment by Shane32** [comment]
Agree
